File: scripts/generate-keymaster.js
#!/usr/bin/env node
/**
 * GGMP Keymaster Generator
 * 
 * Generates custom keymaster keys for GGMP servers
 * This script creates license keys for local/private server deployments
 * 
 * Usage:
 *   node generate-keymaster.js
 *   node generate-keymaster.js --output keys.json
 *   node generate-keymaster.js --count 10
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// Parse command line arguments
const args = process.argv.slice(2);
let outputFile = 'keymaster-keys.json';
let keyCount = 1;
let serverName = 'GGMP Server';

for (let i = 0; i < args.length; i++) {
    if (args[i] === '--output' && i + 1 < args.length) {
        outputFile = args[i + 1];
        i++;
    } else if (args[i] === '--count' && i + 1 < args.length) {
        keyCount = parseInt(args[i + 1]);
        i++;
    } else if (args[i] === '--name' && i + 1 < args.length) {
        serverName = args[i + 1];
        i++;
    } else if (args[i] === '--help' || args[i] === '-h') {
        console.log(`
GGMP Keymaster Generator

Usage:
  node generate-keymaster.js [options]

Options:
  --output <file>    Output file for keys (default: keymaster-keys.json)
  --count <number>   Number of keys to generate (default: 1)
  --name <name>      Server name (default: GGMP Server)
  --help, -h         Show this help message

Examples:
  node generate-keymaster.js
  node generate-keymaster.js --output server-keys.json
  node generate-keymaster.js --count 5 --name "My GGMP Server"
        `);
        process.exit(0);
    }
}

/**
 * Generate a random keymaster key
 */
function generateKey() {
    // Generate a 32-byte random key
    const randomBytes = crypto.randomBytes(32);
    
    // Create a key in the format: GGMP-XXXX-XXXX-XXXX-XXXX-XXXX
    const parts = [];
    for (let i = 0; i < 5; i++) {
        const segment = randomBytes.slice(i * 4, (i + 1) * 4).toString('hex').toUpperCase().substring(0, 4);
        parts.push(segment);
    }
    
    return 'GGMP-' + parts.join('-');
}

/**
 * Generate key metadata
 */
function generateKeyMetadata(key, index) {
    const now = new Date().toISOString();
    
    return {
        key: key,
        id: index + 1,
        serverName: serverName,
        generated: now,
        expiresAt: null, // No expiration for local keys
        maxPlayers: 2048,
        features: {
            customAuth: true,
            enhancedStreaming: true,
            unlimitedPlayers: true,
            premiumPerks: true
        },
        type: 'GGMP_LOCAL',
        version: '1.0.0'
    };
}

/**
 * Main function
 */
function main() {
    console.log('╔══════════════════════════════════════════╗');
    console.log('║   GGMP Keymaster Key Generator v1.0      ║');
    console.log('╚══════════════════════════════════════════╝');
    console.log('');
    
    console.log(`Generating ${keyCount} key(s) for: ${serverName}`);
    console.log('');
    
    const keys = [];
    
    for (let i = 0; i < keyCount; i++) {
        const key = generateKey();
        const metadata = generateKeyMetadata(key, i);
        keys.push(metadata);
        
        console.log(`✓ Generated key ${i + 1}/${keyCount}: ${key}`);
    }
    
    // Save to file
    const output = {
        generator: 'GGMP Keymaster Generator',
        version: '1.0.0',
        generated: new Date().toISOString(),
        serverName: serverName,
        totalKeys: keyCount,
        keys: keys
    };
    
    fs.writeFileSync(outputFile, JSON.stringify(output, null, 2));
    
    console.log('');
    console.log(`✓ Keys saved to: ${outputFile}`);
    console.log('');
    console.log('═══════════════════════════════════════════');
    console.log('Configuration Instructions:');
    console.log('═══════════════════════════════════════════');
    console.log('');
    console.log('Add to your server.cfg:');
    console.log(`  set sv_customLicenseKey "${keys[0].key}"`);
    console.log('');
    console.log('Or use environment variable:');
    console.log(`  export GGMP_LICENSE_KEY="${keys[0].key}"`);
    console.log('');
    console.log('For multiple servers, use different keys from the generated file.');
    console.log('');
}

// Run the generator
try {
    main();
} catch (error) {
    console.error('Error generating keys:', error.message);
    process.exit(1);
}

File: scripts/generate-keymaster.py
#!/usr/bin/env python3
"""
GGMP Keymaster Generator (Python)

Generates custom keymaster keys for GGMP servers
This script creates license keys for local/private server deployments

Usage:
    python generate-keymaster.py
    python generate-keymaster.py --output keys.json
    python generate-keymaster.py --count 10
"""

import os
import sys
import json
import secrets
import argparse
from datetime import datetime


def generate_key():
    """Generate a random keymaster key"""
    # Generate random bytes
    random_bytes = secrets.token_bytes(20)
    
    # Create a key in the format: GGMP-XXXX-XXXX-XXXX-XXXX-XXXX
    parts = []
    for i in range(5):
        segment = random_bytes[i*4:(i+1)*4].hex().upper()[:4]
        parts.append(segment)
    
    return 'GGMP-' + '-'.join(parts)


def generate_key_metadata(key, index, server_name):
    """Generate key metadata"""
    now = datetime.now().astimezone().replace(microsecond=0).isoformat()
    
    return {
        'key': key,
        'id': index + 1,
        'serverName': server_name,
        'generated': now,
        'expiresAt': None,  # No expiration for local keys
        'maxPlayers': 2048,
        'features': {
            'customAuth': True,
            'enhancedStreaming': True,
            'unlimitedPlayers': True,
            'premiumPerks': True
        },
        'type': 'GGMP_LOCAL',
        'version': '1.0.0'
    }


def main():
    """Main function"""
    parser = argparse.ArgumentParser(
        description='GGMP Keymaster Key Generator',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python generate-keymaster.py
  python generate-keymaster.py --output server-keys.json
  python generate-keymaster.py --count 5 --name "My GGMP Server"
        """
    )
    
    parser.add_argument('--output', default='keymaster-keys.json',
                        help='Output file for keys (default: keymaster-keys.json)')
    parser.add_argument('--count', type=int, default=1,
                        help='Number of keys to generate (default: 1)')
    parser.add_argument('--name', default='GGMP Server',
                        help='Server name (default: GGMP Server)')
    
    args = parser.parse_args()
    
    print('╔══════════════════════════════════════════╗')
    print('║   GGMP Keymaster Key Generator v1.0      ║')
    print('╚══════════════════════════════════════════╝')
    print()
    
    print(f'Generating {args.count} key(s) for: {args.name}')
    print()
    
    keys = []
    
    for i in range(args.count):
        key = generate_key()
        metadata = generate_key_metadata(key, i, args.name)
        keys.append(metadata)
        
        print(f'✓ Generated key {i + 1}/{args.count}: {key}')
    
    # Save to file
    output = {
        'generator': 'GGMP Keymaster Generator',
        'version': '1.0.0',
        'generated': datetime.now().astimezone().replace(microsecond=0).isoformat(),
        'serverName': args.name,
        'totalKeys': args.count,
        'keys': keys
    }
    
    with open(args.output, 'w') as f:
        json.dump(output, f, indent=2)
    
    print()
    print(f'✓ Keys saved to: {args.output}')
    print()
    print('═══════════════════════════════════════════')
    print('Configuration Instructions:')
    print('═══════════════════════════════════════════')
    print()
    print('Add to your server.cfg:')
    print(f'  set sv_customLicenseKey "{keys[0]["key"]}"')
    print()
    print('Or use environment variable:')
    print(f'  export GGMP_LICENSE_KEY="{keys[0]["key"]}"')
    print()
    print('For multiple servers, use different keys from the generated file.')
    print()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('\n\nAborted by user.')
        sys.exit(1)
    except Exception as e:
        print(f'Error generating keys: {e}', file=sys.stderr)
        sys.exit(1)
